src/preprocessing/passes/bv_ackermann.cpp

```diff
@@ -54,42 +94,82 @@ void storeFunction(
   NodeSet& set = fun_to_args[func];
   if (set.find(term) == set.end())
   {
-    set.insert(term);
     TypeNode tn = term.getType();
-    Node skolem = NodeManager::currentNM()->mkSkolem(
+    Node skolem = nm->mkSkolem(
         "BVSKOLEM$$",
         tn,
         "is a variable created by the ackermannization "
         "preprocessing pass for theory BV");
+    for (const auto& t : set)
+    {
+      addLemmaForPair(t, term, func, assertions, nm, stack);
+    }
+    set.insert(term);
     fun_to_skolem.addSubstitution(term, skolem);
+    /* add the arguments to the stack so that
+     * collectFunctionsAndLemmas will process them as well. */
+    for (TNode arg : term)
```

Comment:
A slightly tricky optimization can be added: if set has only one term in it, we don't need to add its arguments to the stack.  Of course, then, as soon as it has two terms, we have to add args from both terms to the stack.